src/modules/persistent-store-base.ts

```diff
@@ -0,0 +1,76 @@
+import { Store } from 'store-unit';
+
+type Options<S> = {
+  retrieve: (key: string) => Promise<S | undefined>;
+  save: (key: string, value: S) => Promise<void>;
+};
+
+export class PersistentStoreBase<T> extends Store<T> {
+  protected key: string;
+  protected isReady: boolean;
+  private readyPromise: Promise<void>;
+
+  public options: Options<T>;
+  public defaultOptions: Options<T> = {
+    retrieve: () => {
+      throw new Error('PersistentStore:retrieve Not Implemented');
+    },
+    save: () => {
+      throw new Error('PersistentStore:save Not Implemented');
+    },
+  };
+
+  constructor(initialState: T, key: string, options: Partial<Options<T>> = {}) {
+    super(initialState);
+    this.key = key;
+    this.options = { ...this.defaultOptions, ...options };
+    this.isReady = false;
+    this.readyPromise = this.restore();
+    this.on('change', (state) => {
+      // only write to disk after restore so that we're not making
+      // a redundant write on each initialization
+      if (this.isReady) {
+        this.options.save(this.key, state);
+      }
+    });
+  }
+
+  getState() {
+    if (!this.isReady) {
+      throw new Error('Do not access getState() before checking ready()');
+    }
+    return super.getState();
+  }
+
+  setState(...args: Parameters<Store<T>['setState']>) {
+    if (!this.isReady) {
+      if (process.env.NODE_ENV === 'development') {
+        // Throw only in dev mode in case some production flow depends on
+        // setting state sooner. Before this refactoring it was possible, so
+        // it shouldn't really break anything
+        throw new Error(
+          'You are trying to write to a PersistentStore before {ready}'
+        );
+      }
+    }
+    super.setState(...args);
```

Comment:
As I can see here, we have both sync and async state updates. Can this cause the racecondition when sync updates will be called in one order, but async saves will be executed in another order which will cause incorrect state after store reload?